proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,200 @@
+
+```csharp
+var f1 = () => default;        // error: no natural type
+var f2 = x => { };             // error: no natural type
+var f3 = x => x;               // error: no natural type
+var f4 = () => 1;              // System.Func<int>
+var f5 = () : string => null;  // System.Func<string>
+```
+
+```csharp
+static void F1() { }
+static void F1<T>(this T t) { }
+static void F2(this string s) { }
+
+var f6 = F1;    // error: multiple methods
+var f7 = "".F1; // System.Action
+var f8 = F2;    // System.Action<string> 
+```
+
+The synthesized delegate types are implicitly co- and contra-variant.
+```csharp
+var fA = (IEnumerable<string> e, ref int i) => { }; // void DA$(IEnumerable<string>, ref int);
+fA = (IEnumerable<object> e, ref int i) => { };     // ok
+
+var fB = (IEnumerable<object> e, ref int i) => { }; // void DB$(IEnumerable<object>, ref int);
+fB = (IEnumerable<string> e, ref int i) => { };     // error: parameter type mismatch
+```
+
+### Implicit conversion to `System.Delegate`
+A consequence of inferring a natural type is that lambda expressions and method groups with natural type are implicitly convertible to `System.Delegate`.
+```csharp
+static void Invoke(Func<string> f) { }
+static void Invoke(Delegate d) { }
+
+static string GetString() => "";
+static int GetInt() => 0;
+
+Invoke(() => "");  // Invoke(Func<string>)
+Invoke(() => 0);   // Invoke(Delegate) [new]
+
+Invoke(GetString); // Invoke(Func<string>)
+Invoke(GetInt);    // Invoke(Delegate) [new]
+```
+
+If a natural type cannot be inferred, there is no implicit conversion to `System.Delegate`.
+```csharp
+static void Invoke(Delegate d) { }
+
+Invoke(Console.WriteLine); // error: cannot to 'Delegate'; multiple candidate methods
+Invoke(x => x);            // error: cannot to 'Delegate'; no natural type for 'x'
+```
+
+To avoid a breaking change, overload resolution will be updated to prefer strongly-typed delegates and expressions over `System.Delegate`.
+_The example below demonstrates the tie-breaking rule for lambdas. Is there an equivalent example for method groups?_
+```csharp
+static void Execute(Expression<Func<string>> e) { }
+static void Execute(Delegate d) { }
+
+static string GetString() => "";
+static int GetInt() => 0;
+
+Execute(() => "");  // Execute(Expression<Func<string>>) [tie-breaker]
+Execute(() => 0);   // Execute(Delegate) [new]
+
+Execute(GetString); // Execute(Delegate) [new]
+Execute(GetInt);    // Execute(Delegate) [new]
+```
+
+## Syntax
+
+```antlr
+lambda_expression
+  : attribute_list* modifier* lambda_parameters (':' type)? '=>' (block | body)
+  ;
+
+lambda_parameters
+  : lambda_parameter
+  | '(' (lambda_parameter (',' lambda_parameter)*)? ')'
+  ;
+
+lambda_parameter
+  : identifier
+  | (attribute_list* modifier* type)? identifier equals_value_clause?
+  ;
+```
+
+_Does the `: type` return type syntax introduce ambiguities with `?:` that cannot be resolved easily?_
```

Comment:
Consider including the alternative forms we discussed in LDM.

Reply:
I'll update the syntax based on the LDM and subsequent meetings in a separate PR.

---
In reply to: [588782316](https://github.com/dotnet/csharplang/pull/4451#discussion_r588782316) [](ancestors = 588782316)